agent/rpcclient/health/health.go

```diff
@@ -0,0 +1,64 @@
+package health
+
+import (
+	"context"
+
+	"github.com/hashicorp/consul/agent/cache"
+	cachetype "github.com/hashicorp/consul/agent/cache-types"
+	"github.com/hashicorp/consul/agent/structs"
+)
+
+type Client struct {
+	NetRPC NetRPC
+	Cache  CacheGetter
+}
+
+type NetRPC interface {
+	RPC(method string, args interface{}, reply interface{}) error
+}
+
+type CacheGetter interface {
+	Get(ctx context.Context, t string, r cache.Request) (interface{}, cache.ResultMeta, error)
+}
+
+func (c *Client) ServiceNodes(
+	ctx context.Context,
+	req structs.ServiceSpecificRequest,
+) (structs.IndexedCheckServiceNodes, cache.ResultMeta, error) {
+	out, md, err := c.getServiceNodes(ctx, req)
+	if err != nil {
+		return out, md, err
+	}
+
+	// TODO: DNSServer emitted a metric here, do we still need it?
```

Comment:
The metric that was emitted here was:

```go
	if args.AllowStale && out.LastContact > staleCounterThreshold {	
		metrics.IncrCounter([]string{"dns", "stale_queries"}, 1)	
	}
```

From what I could tell this was added a few years ago as a way of turning the `dns_config.MaxStale` argument. Do we know if this query is still actively used?

I think there are a couple options for restoring this. 

1. Replace the `cache.ResultMeta` return parameter with a `Metadata` that includes details about the first call. That would allow us to emit the metric from the DNS server after this method returns. The disadvantage to that approach is that we'll likely need to put that Metadata type into some other package so that it can be used by other implementations.

2. Alternatively we could have the `Client` accept something that emits the metric, and we could pass in different instances of `Client` for dns and httpapi.

I'm not sure yet which one of those would work better going forward, so I thought I would ask about the metric first before implementing either of those.

Reply:
Why does the metric have to change at all. The `IndexedCheckServiceNodes` still contains the last contact time in it which would allow it to continue to emit the metric if it wanted.

Whether it is a useful metric is another question as is whether the previous implementation was correct. Our docs say this about the metric:

> This increments when an agent serves a query within the allowed stale threshold.

Which isn't exactly true. This metric increments for any stale query fielded by an agent whose last contact time is more than 5s in the past.
 
What would be more useful would be to have a metric which reported on the average query staleness. We could call the `AddSample` method from armon/go-metrics for each to figure out the min/mean/max/count of stale queries. Whether we should filter out non-stale ones (`LastContact == 0`) I am not sure. What I am thinking though is that the metric as exists today isn't very useful.

Reply:
> Why does the metric have to change at all. The IndexedCheckServiceNodes still contains the last contact time in it which would allow it to continue to emit the metric if it wanted.

The rpcclient implementation does up to 2 RPC calls. That LastContact time could be for the second RPC call, and in many cases where the metric would have fired I think we would also perform 2 RPC calls. Previously the metric always looked at the LastContact for the first RPC call. Sine the second call uses `AllowStale` false, I presume that means the LastContact time should be 0.

Reply:
> The rpcclient implementation does up to 2 RPC calls.

This seems like a symptom of another problem which I never noticed before but this makes blatantly obvious.

Currently we parse the max staleness values into the request structs in the HTTP API and use gotos to retry requests as not stale if it fails. A smallish example is here: https://github.com/hashicorp/consul/blob/bc5ca8b6384f87ee5edc8949e18d2afad0941373/agent/catalog_endpoint.go#L114-L139

The problem here is that the max stale duration value is in the request struct that we send along to the current server in the RPC. The server then calls `ForwardRPC` which checks if it should handle the RPC for forward to the leader/another DC. Its at that point in time where we should check the max stale duration against the `LastContact` time. Instead we just check if stale queries are allowed and if so handle it ignoring any duration value in the request.

Not only could moving the checks to the central location simplify a lot of code it could actually make bounded staleness in the ACL, config entry and intentions HTTP APIs actually work.

This is probably outside of scope for this PR though.